codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/compression/RequestCompressionIntegration.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.smithy.kotlin.codegen.rendering.compression
+
+import software.amazon.smithy.kotlin.codegen.KotlinSettings
+import software.amazon.smithy.kotlin.codegen.core.CodegenContext
+import software.amazon.smithy.kotlin.codegen.core.KotlinWriter
+import software.amazon.smithy.kotlin.codegen.core.RuntimeTypes
+import software.amazon.smithy.kotlin.codegen.core.withBlock
+import software.amazon.smithy.kotlin.codegen.integration.KotlinIntegration
+import software.amazon.smithy.kotlin.codegen.model.defaultValue
+import software.amazon.smithy.kotlin.codegen.model.getTrait
+import software.amazon.smithy.kotlin.codegen.model.hasTrait
+import software.amazon.smithy.kotlin.codegen.model.nonNullable
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolMiddleware
+import software.amazon.smithy.kotlin.codegen.rendering.util.ConfigProperty
+import software.amazon.smithy.kotlin.codegen.rendering.util.ConfigPropertyType
+import software.amazon.smithy.kotlin.codegen.rendering.util.nestedBuilder
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.traits.RequestCompressionTrait
+
+/**
+ * Adds request compression settings if request compression trait is present.
+ * Adds middleware that registers interceptor and provides functionality for request compression if request compression
+ * is not disabled via settings.
+ */
+class RequestCompressionIntegration : KotlinIntegration {
+
+    override fun enabledForService(model: Model, settings: KotlinSettings): Boolean =
+        model.isTraitApplied(RequestCompressionTrait::class.java)
+
+    override fun customizeMiddleware(
+        ctx: ProtocolGenerator.GenerationContext,
+        resolved: List<ProtocolMiddleware>,
+    ): List<ProtocolMiddleware> = resolved + requestCompressionTraitMiddleware
+
+    override fun additionalServiceConfigProps(ctx: CodegenContext): List<ConfigProperty> =
+        listOf(
+            ConfigProperty {
+                name = "requestCompression"
+                symbol = RuntimeTypes.SmithyClient.Config.RequestCompressionConfig
+                builderSymbol = RuntimeTypes.SmithyClient.Config.RequestCompressionConfig.nestedBuilder.toBuilder()
```

Comment:
**correctness**: This config property doesn't look like the other properties we've defined DSL's for. I can't say it's wrong at a glance but take a look at [RetryStrategy](https://github.com/awslabs/smithy-kotlin/blob/f5feb5638394010118a5234b450bd4020920871c/codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/util/RuntimeConfigProperty.kt#L90) and [HttpClient](https://github.com/awslabs/smithy-kotlin/blob/f5feb5638394010118a5234b450bd4020920871c/codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/util/RuntimeConfigProperty.kt#L43) to see if this can't be simplified.